docs/training/apis-howto/api-core-ug.rst

```diff
@@ -1,212 +1,302 @@
+our script. More specifically, we'll create a :class:`~determined.core.Context` object to allow
+interaction with the master. Then, we'll pass the ``core_context`` as an argument into ``main()``,
+``train()``, and ``test()`` and modify the function headers accordingly.
 
-#. Within ``main()``, add two calls: (1) report training metrics periodically during training and
-   (2) report validation metrics every time a validation runs.
+To run our experiment, we'll use the ``model_def_metrics.py`` script and its accompanying
+``metrics.yaml`` experiment configuration file.
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/1_metrics.py
-      :language: python
-      :pyobject: main
+Begin by importing Determined:
 
-   The ``report_validation_metrics()`` call typically happens after the validation step, however,
-   actual validation is not demonstrated by this example.
+.. code:: python
 
-#. Create a ``1_metrics.yaml`` file with an ``entrypoint`` invoking the new ``1_metrics.py`` file.
-   You can copy the ``0_start.yaml`` configuration file and change the first couple of lines:
+   import determined as det
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/1_metrics.yaml
-      :language: yaml
-      :lines: 1-2
+Step 2.1: Modify the Main Loop
+==============================
 
-#. Run the code using the command:
+We'll need a ``core.Context`` object for interacting with the master. To accomplish this, we'll
+modify the __main__loop to include ``core_context``:
 
-   .. code:: bash
+.. note::
 
-      det e create 1_metrics.yaml . -f
+   Refer to the ``if __name__ == "__main__":`` block in ``model_def_metrics.py``
 
-#. You can now navigate to the new experiment in the WebUI and view the plot populated with the
-   training and validation metrics.
+.. literalinclude:: ../../../examples/tutorials/core_api_pytorch_mnist/model_def_metrics.py
+   :language: python
+   :start-after: # Docs snippet start: modify main loop core context
+   :end-at: main(core_context=core_context)
 
-The complete ``1_metrics.py`` and ``1_metrics.yaml`` listings used in this example can be found in
-the :download:`core_api.tgz </examples/core_api.tgz>` download or in the `Github repository
-<https://github.com/determined-ai/determined/tree/master/examples/tutorials/core_api>`_.
+Step 2.2: Modify the Train Method
+=================================
 
-.. _core-checkpoints:
+Use ``core_context.train`` to report training and validation metrics.
 
-********************
- Report Checkpoints
-********************
+Modify the train() method by adding ``core_context.train.report_training_metrics()``:
 
-By checkpointing periodically during training and reporting those checkpoints to the master, you can
-stop and restart training in two different ways: either by pausing and reactivating training using
-the WebUI, or by clicking the Continue Trial button after the experiment completes.
```

Comment:
These side by side `report_training_metrics` and `report_validation_metrics` are pretty confusing, IMO.  Because of a lot of unnecessary assymetry:
* In the first block, we have some long computation for `steps_completed`: `steps_completed=batches_completed + epoch_idx * len(train_loader)` while in the second it's just `steps_completed=steps_completed`
* For the two losses, calling `.item()` for the `train_loss` value, but not when computing `test_loss`
* Using `epoch_idx` for the first `epoch` value, but just `epoch` for the second.

Looks like the referenced code just needs cleanup.

Reply:
@rb-determined-ai please comment on this

Reply:
epochs_completed (count-like things) or epoch_idx (index like things), not epoch

remove empty lines before the end ")"

Reply:
I agree with @garrett361 that some work on the example is needed, and probably should not fall on @tara-det-ai.

I do think at least in this case, that we could make a separate ticket and it doesn't need to block this PR.

Reply:
Agreed.

Reply:
DOCSTW-175